### ISSUE TITLE
reportOverlappingOverload false positive: variadic tuple unpacking

### ISSUE THREAD

**Issue by ringohoffman** [microsoft/pyright#7987]
**Describe the bug**
I am getting `reportOverlappingOverload` for 2 signatures that do not overlap.

**Code or Screenshots**

I am trying to add some more expressive type hints for PyTorch Lightning's [`Fabric.setup()`](https://github.com/Lightning-AI/pytorch-lightning/blob/896c2a656ad2db3278ec11520aed04e378f4462b/src/lightning/fabric/fabric.py#L211-L217).

There are 2 cases:

1. no optimizers -> return just `_FabricModule`.
2. at least one optimizer -> return tuple of `_FabricModule` and the same number of `_FabricOptimizer` as the provided number of `Optimizer` (sadly I don't think I can use `TypeVarTuple` here to help validate the number of `_FabricOptimizer` returned).

The return values are typed correctly, but I am getting a false positive for `reportOverlappingOverload`. `mypy` does not detect an overlapping signature either.

```python
from typing import overload, Union, Tuple

from typing_extensions import Unpack


class Module:
    ...

class Linear(Module):
    ...

class Optimizer:
    ...

class Adam(Optimizer):
    ...

class _FabricModule:
    ...

class _FabricOptimizer:
    ...


@overload
def setup(  # Overload 1 for "setup" overlaps overload 2 and returns an incompatible type (reportOverlappingOverload)
    module: Module,
    *optimizers: Unpack[Tuple[()]],
    move_to_device: bool = ...,
    _reapply_compile: bool = ...,
) -> _FabricModule: ...

@overload
def setup(
    module: Module,
    *optimizers: Unpack[Tuple[Optimizer, Unpack[Tuple[Optimizer, ...]]]],
    move_to_device: bool = ...,
    _reapply_compile: bool = ...,
) -> Tuple[_FabricModule, Unpack[Tuple[_FabricOptimizer, ...]]]: ...

def setup(
    module: Module,
    *optimizers: Optimizer,
    move_to_device: bool = True,
    _reapply_compile: bool = True,
) -> Union[_FabricModule, Tuple[_FabricModule, Unpack[Tuple[_FabricOptimizer, ...]]]]:
    ...


model = Linear()
opt = Adam()

wrapped_model = setup(model)
reveal_type(wrapped_model)  # Type of "wrapped_model" is "_FabricModule"

wrapped_model, wrapped_opt = setup(model, opt)
reveal_type(wrapped_model)  # Type of "wrapped_model" is "_FabricModule"
reveal_type(wrapped_opt)  # Type of "wrapped_opt" is "_FabricOptimizer"

wrapped_model, wrapped_opt1, wrapped_opt2 = setup(model, opt, opt)
reveal_type(wrapped_model)  # Type of "wrapped_model" is "_FabricModule"
reveal_type(wrapped_opt1)  # Type of "wrapped_opt1" is "_FabricOptimizer"
reveal_type(wrapped_opt2)  # Type of "wrapped_opt2" is "_FabricOptimizer"
```

If you create 2 separate functions matching each of the overload signature, you can see they don't overlap:

```python
# like overload 1
def no_optimizers(*optimizer: Unpack[Tuple[()]]) -> None:
    ...

# like overload 2
def at_least_one_optimizer(*optimizers: Unpack[Tuple[Optimizer, Unpack[Tuple[Optimizer, ...]]]]) -> None:
    ...


model = Linear()
opt = Adam()

no_optimizers()
no_optimizers(opt)  # Expected 0 positional arguments
no_optimizers(opt, opt)  # Expected 0 positional arguments

at_least_one_optimizer()  # Argument missing for parameter "optimizers[0]"
at_least_one_optimizer(opt)
at_least_one_optimizer(opt, opt)
```

[pyright playground link](https://pyright-play.net/?pythonVersion=3.13&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgDRQCqKSYK9AKgK4JnEBQfUJFiJUaAPrEAHjGIoAziwWZsufEwSUAxgGsBfLWUrz5UALJhqnXgC4%2BUB1AB0LgYeOmAMqmKUQACgsrXgBKO0dnVwMjEygAeQQYLCQAL1JwxxcnNxjTAEFqSgh-BKSIVNIw%2B0yo91jxADFKACMQJC0g62IMhyycjyhGlratUuS0kB7I7IEAASJSCho%2BamJgKHliGG5-aocISy6bc0PeWj2oACowRPHSeWONbR0AbS4eYhf-EIBdH-OIgcSOIYGBxKtCO1ulBmmAwGQoABeaYAxziEC%2BBA8ODiLSQZC2GFwhHIrLnEJQAC0AD5Bk1Wu1OoS%2Bnx5iRyFRqCs1hstjsLgdgtCmcRUQ5rrdyhMHowUJpdG9uLwXmMpaR6E8Fe9laqKiB6Fk-n8xVAgcQQWCIVDjrD4UiURd0Zjsbj8UhCbaSQ6KTSoNrPkMGR1TqLZfLXv6XoGRrqJgaXEbjizVutNtsELtASHjiKTRKynqZbH1QKFhbwcRIVpoZ77ewQJxRY6MZQsWQcXjsO6a8S6w2mz7aUwlFH6SNc36lQGx4yQxq5c9FR9R8N2sX9dMjT8piy%2BABiKBkJA6YgEdlLahQACM3PWKDBNwL0v8%2BbukzDi8j3z%2Bg6gADlWG6C4WQPI8TzPRZOSgAAmW8oEoGBxF4YxEMA8RHzfF8MLVEAZU1CMpxVSU9XncMlx1Yi403I1fwAlAgIiXdBWIL1vHovxvj4R97QKIpOL4AB3EBWwQYhqHEZivTTHZJJCPgMRISgyBBRBiH8ISRLEiTLBYuTBOErEtMk%2BgNMM8TuORaSM2MghEjkhTfGU%2BBRPUgzRPE2T5MrRyVJc0z3PQuyBH8oydLIEy3K0x8rwizTzMSaD7Ss-wbMfehH3s7ylN8tSQo8sLMsUpzVNcuLApgK9Cp85zcsi%2BKYGgvT73Kt95E45rsMLfwMr4DrKPubrEnSoK%2BAQpDfHkVD6JanDOLG5DJvQ6bOomQaYDk%2BaJqm80VtINbhvWvggA)

**VS Code extension or command-line**

```console
$ pyright --version
pyright 1.1.364
```

**Comment by erictraut** [comment]
Thanks for the bug report.

Here's a minimal repro:

```python
@overload
def setup() -> str: ...

@overload
def setup(*args: *tuple[int, *tuple[int, ...]]) -> int: ...
```

Note that the parameter `*optimizers: Unpack[Tuple[()]]` in your code sample can be replaced with `*`. It's a verbose way of saying "no additional positional arguments".

**Comment by erictraut** [comment]
This is included in pyright 1.1.365.
